### ISSUE TITLE
Numeric field queries no longer support large numbers

### ISSUE THREAD

**Issue by rthadr** [RedisLabsModules/RediSearch#134]
Hey all,

Thanks for the work on this awesome plugin.  I found an issue with querying by numeric fields with large values in the latest release.  I'm assuming any release after 0.19.3 is affected.  

Here are steps to reproduce:
```
127.0.0.1:6379> ft.create idx:test schema num numeric
OK
127.0.0.1:6379> ft.add idx:test test:1 1 fields num 89999800
OK
127.0.0.1:6379> ft.add idx:test test:2 1 fields num 89999801
OK
127.0.0.1:6379> ft.search idx:test "@num:[89999800 89999800]"
1) (integer) 2
2) "test:2"
3) 1) "num"
   2) "89999801"
4) "test:1"
5) 1) "num"
   2) "89999800"
127.0.0.1:6379> ft.search idx:test "@num:[89999801 89999801]"
1) (integer) 0
```

Sending the first query, I'd expect to get "test:1" only.  Sending the second, I expect to get "test:2", but no results are returned.  

In versions prior to 0.19.4 numbers of even 10 or 11 digits worked as expected.

BTW, I'm using Redis 4.0.1.

**Comment by dvirsky** [comment]
We've switched to using 32 bit float instead of double in internal encoding, for memory saving. My bad for not documenting the change.

I suppose we can use smarter encoding based on the value at hand. @mnunberg care to take this one? 

There are two paths here: simpler one is to create a flag for numeric fields indicating double precision, i.e. `FT.CREAT ... num NUMERIC [DOUBLE]`. 

Second one - a bit trickier to implement: per numeric record try to encode the value as float if possible, and revert to double if not. The only question is how to encode this internally per record without losing too much RAM and CPU for decoders. 

WDYT?

**Comment by dvirsky** [comment]
Re the second option - we could break the double into sign, mantissa and exponent, and encode them as three varints. it will slow decoding but will actually save more space with some float values.

**Comment by rthadr** [comment]
I have a work around (thanks to reading your reply above).  Since I am dealing with integers here, I realized can simply pre-pend a decimal in front of the number (essentially making it the mantissa) during storage and query.  I verified this workaround does work for my usecase.

**Comment by dvirsky** [comment]
Thanks, I'm leaning towards keeping the internal value a double and being smart internally about the actual encoding. We'll sleep on it a bit if it's not urgent to you ;)

**Comment by dvirsky** [comment]
@rthadr we'll have a fix in the next release (and in master probably today). basically as specified above, internally we'll use doubles, and in the encoded index we'll use some heuristics to store the numbers in the most efficient way based on their properties. So there won't be a need for your workaround and we'll get massive memory gains in most cases - in your case since the numbers are actually integers the encoder will detect that and use compact integer compression.

**Comment by rthadr** [comment]
That is great!  I look forward to it.

**Comment by mnunberg** [comment]
@rthadr https://github.com/RedisLabsModules/RediSearch/pull/137 - not yet merged, but you can try it out and see if this works for you.

**Comment by rthadr** [comment]
Thanks for this.  Getting a crash when rapidly indexing a batch of large integers:
```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
1:M 18 Aug 03:12:28.953 # Redis 4.0.1 crashed by signal: 11
1:M 18 Aug 03:12:28.953 # Crashed running the instuction at: 0x7ff0bdd90f59
1:M 18 Aug 03:12:28.953 # Accessing address: 0x7ff18acfabaf
1:M 18 Aug 03:12:28.953 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
EIP:
/usr/lib/redisearch.so(WriteVarint+0x49)[0x7ff0bdd90f59]

Backtrace:
redis-server *:6379(logStackTrace+0x29)[0x467a89]
redis-server *:6379(sigsegvHandler+0xa6)[0x468126]
/lib/x86_64-linux-gnu/libpthread.so.0(+0xf890)[0x7ff0c0130890]
/usr/lib/redisearch.so(WriteVarint+0x49)[0x7ff0bdd90f59]

------ INFO OUTPUT ------
# Server
redis_version:4.0.1
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:de1f157009f8c4af
redis_mode:standalone
os:Linux 4.9.30-moby x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:4.9.2
process_id:1
run_id:6f9153b4550ff58af9e78a4910e268ecee48fbab
tcp_port:6379
uptime_in_seconds:11
uptime_in_days:0
hz:10
lru_clock:9853724
executable:/data/redis-server
config_file:

# Clients
connected_clients:1
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:1

# Memory
used_memory:26442568
used_memory_human:25.22M
used_memory_rss:15097856
used_memory_rss_human:14.40M
used_memory_peak:26442568
used_memory_peak_human:25.22M
used_memory_peak_perc:102.46%
used_memory_overhead:877558
used_memory_startup:765584
used_memory_dataset:25565010
used_memory_dataset_perc:99.56%
total_system_memory:2076463104
total_system_memory_human:1.93G
used_memory_lua:37888
used_memory_lua_human:37.00K
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
mem_fragmentation_ratio:0.57
mem_allocator:jemalloc-4.0.3
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:1141
rdb_bgsave_in_progress:0
rdb_last_save_time:1503025937
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:1
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
aof_current_size:277055
aof_base_size:0
aof_pending_rewrite:0
aof_buffer_length:0
aof_rewrite_buffer_length:0
aof_pending_bio_fsync:0
aof_delayed_fsync:0

# Stats
total_connections_received:1
total_commands_processed:3426
instantaneous_ops_per_sec:1369
total_net_input_bytes:360691
total_net_output_bytes:14223
instantaneous_input_kbps:142.72
instantaneous_output_kbps:4.46
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
evicted_keys:0
keyspace_hits:1141
keyspace_misses:1
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:4b6fb5a3ae05aed24fa6e6011069bb397f083734
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:0.18
used_cpu_user:0.15
used_cpu_sys_children:0.00
used_cpu_user_children:0.00

# Commandstats
cmdstat_hmset:calls=1141,usec=10168,usec_per_call=8.91
cmdstat_hgetall:calls=1141,usec=4410,usec_per_call=3.87
cmdstat_info:calls=1,usec=96,usec_per_call=96.00

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=1149,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=2 addr=172.17.0.1:44702 fd=9 name= age=3 idle=0 flags=b db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=0 oll=0 omem=0 events=r cmd=FT.ADDHASH

------ REGISTERS ------
1:M 18 Aug 03:12:28.954 # 
RAX:00000000ffffffff RBX:0000000000000001
RCX:00000000ffffffff RDX:00007ff18acfabaf
RDI:00000000ffffffff RSI:00007ff08acfac60
RBP:00007ff08acfac60 RSP:00007ff08acfabb0
R8 :00000000fffffffe R9 :00007ff08acfabb0
R10:fffffffffffe9600 R11:0000000000000015
R12:00007ff08acfac60 R13:00000000000000eb
R14:000000a1a81ff000 R15:0000000000000005
RIP:00007ff0bdd90f59 EFL:0000000000010286
CSGSFS:002b000000000033
1:M 18 Aug 03:12:28.954 # (00007ff08acfabbf) -> 00007ff08a2bf4b8
1:M 18 Aug 03:12:28.954 # (00007ff08acfabbe) -> 0000000000000010
1:M 18 Aug 03:12:28.954 # (00007ff08acfabbd) -> 00007ff0bdd8c7d4
1:M 18 Aug 03:12:28.954 # (00007ff08acfabbc) -> c2643503fe000000
1:M 18 Aug 03:12:28.954 # (00007ff08acfabbb) -> 00007ff08a00ac90
1:M 18 Aug 03:12:28.954 # (00007ff08acfabba) -> 0000000000000001
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb9) -> 0000000000000001
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb8) -> 00007ff07c00eca0
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb7) -> 00007ff0bdd8c7bf
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb6) -> 00000000000000eb
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb5) -> 00007ff08acfac60
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb4) -> 00000000f445b780
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb3) -> 0000000000000001
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb2) -> 0000000000000381
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb1) -> 00ee95a1fefefefe
1:M 18 Aug 03:12:28.954 # (00007ff08acfabb0) -> fefefefefefefefe

------ FAST MEMORY TEST ------
1:M 18 Aug 03:12:28.954 # Bio thread for job type #0 terminated
1:M 18 Aug 03:12:28.954 # Bio thread for job type #1 terminated
1:M 18 Aug 03:12:28.955 # Bio thread for job type #2 terminated
*** Preparing to test memory region 752000 (94208 bytes)
*** Preparing to test memory region 205c000 (1757184 bytes)
*** Preparing to test memory region 7ff07c000000 (135168 bytes)
*** Preparing to test memory region 7ff080000000 (139264 bytes)
*** Preparing to test memory region 7ff084000000 (143360 bytes)
*** Preparing to test memory region 7ff088000000 (37748736 bytes)
*** Preparing to test memory region 7ff08a4fe000 (8388608 bytes)
*** Preparing to test memory region 7ff08acff000 (8388608 bytes)
*** Preparing to test memory region 7ff08b500000 (8388608 bytes)
*** Preparing to test memory region 7ff08bd01000 (8388608 bytes)
*** Preparing to test memory region 7ff08c502000 (8388608 bytes)
*** Preparing to test memory region 7ff08cd03000 (8388608 bytes)
*** Preparing to test memory region 7ff08d504000 (8388608 bytes)
*** Preparing to test memory region 7ff08dd05000 (8388608 bytes)
*** Preparing to test memory region 7ff08e506000 (8388608 bytes)
*** Preparing to test memory region 7ff08ed07000 (8388608 bytes)
*** Preparing to test memory region 7ff08f508000 (8388608 bytes)
*** Preparing to test memory region 7ff08fd09000 (8388608 bytes)
*** Preparing to test memory region 7ff09050a000 (8388608 bytes)
*** Preparing to test memory region 7ff090d0b000 (8388608 bytes)
*** Preparing to test memory region 7ff09150c000 (8388608 bytes)
*** Preparing to test memory region 7ff091d0d000 (8388608 bytes)
*** Preparing to test memory region 7ff09250e000 (8388608 bytes)
*** Preparing to test memory region 7ff092d0f000 (8388608 bytes)
*** Preparing to test memory region 7ff093510000 (8388608 bytes)
*** Preparing to test memory region 7ff093d11000 (8388608 bytes)
*** Preparing to test memory region 7ff094512000 (8388608 bytes)
*** Preparing to test memory region 7ff094d13000 (8388608 bytes)
*** Preparing to test memory region 7ff095514000 (8388608 bytes)
*** Preparing to test memory region 7ff095d15000 (8388608 bytes)
*** Preparing to test memory region 7ff096516000 (8388608 bytes)
*** Preparing to test memory region 7ff096d17000 (8388608 bytes)
*** Preparing to test memory region 7ff097518000 (8388608 bytes)
*** Preparing to test memory region 7ff097d19000 (8388608 bytes)
*** Preparing to test memory region 7ff09851a000 (8388608 bytes)
*** Preparing to test memory region 7ff098d1b000 (8388608 bytes)
*** Preparing to test memory region 7ff09951c000 (8388608 bytes)
*** Preparing to test memory region 7ff099d1d000 (8388608 bytes)
*** Preparing to test memory region 7ff09a51e000 (8388608 bytes)
*** Preparing to test memory region 7ff09ad1f000 (8388608 bytes)
*** Preparing to test memory region 7ff09b520000 (8388608 bytes)
*** Preparing to test memory region 7ff09bd21000 (8388608 bytes)
*** Preparing to test memory region 7ff09c522000 (8388608 bytes)
*** Preparing to test memory region 7ff09cd23000 (8388608 bytes)
*** Preparing to test memory region 7ff09d524000 (8388608 bytes)
*** Preparing to test memory region 7ff09dd25000 (8388608 bytes)
*** Preparing to test memory region 7ff09e526000 (8388608 bytes)
*** Preparing to test memory region 7ff09ed27000 (8388608 bytes)
*** Preparing to test memory region 7ff09f528000 (8388608 bytes)
*** Preparing to test memory region 7ff09fd29000 (8388608 bytes)
*** Preparing to test memory region 7ff0a052a000 (8388608 bytes)
*** Preparing to test memory region 7ff0a0d2b000 (8388608 bytes)
*** Preparing to test memory region 7ff0a152c000 (8388608 bytes)
*** Preparing to test memory region 7ff0a1d2d000 (8388608 bytes)
*** Preparing to test memory region 7ff0a252e000 (8388608 bytes)
*** Preparing to test memory region 7ff0a2d2f000 (8388608 bytes)
*** Preparing to test memory region 7ff0a3530000 (8388608 bytes)
*** Preparing to test memory region 7ff0a3d31000 (8388608 bytes)
*** Preparing to test memory region 7ff0a4532000 (8388608 bytes)
*** Preparing to test memory region 7ff0a4d33000 (8388608 bytes)
*** Preparing to test memory region 7ff0a5534000 (8388608 bytes)
*** Preparing to test memory region 7ff0a5d35000 (8388608 bytes)
*** Preparing to test memory region 7ff0a6536000 (8388608 bytes)
*** Preparing to test memory region 7ff0a6d37000 (8388608 bytes)
*** Preparing to test memory region 7ff0a7538000 (8388608 bytes)
*** Preparing to test memory region 7ff0a7d39000 (8388608 bytes)
*** Preparing to test memory region 7ff0a853a000 (8388608 bytes)
*** Preparing to test memory region 7ff0a8d3b000 (8388608 bytes)
*** Preparing to test memory region 7ff0a953c000 (8388608 bytes)
*** Preparing to test memory region 7ff0a9d3d000 (8388608 bytes)
*** Preparing to test memory region 7ff0aa53e000 (8388608 bytes)
*** Preparing to test memory region 7ff0aad3f000 (8388608 bytes)
*** Preparing to test memory region 7ff0ab540000 (8388608 bytes)
*** Preparing to test memory region 7ff0abd41000 (8388608 bytes)
*** Preparing to test memory region 7ff0ac542000 (8388608 bytes)
*** Preparing to test memory region 7ff0acd43000 (8388608 bytes)
*** Preparing to test memory region 7ff0ad544000 (8388608 bytes)
*** Preparing to test memory region 7ff0add45000 (8388608 bytes)
*** Preparing to test memory region 7ff0ae546000 (8388608 bytes)
*** Preparing to test memory region 7ff0aed47000 (8388608 bytes)
*** Preparing to test memory region 7ff0af548000 (8388608 bytes)
*** Preparing to test memory region 7ff0afd49000 (8388608 bytes)
*** Preparing to test memory region 7ff0b054a000 (8388608 bytes)
*** Preparing to test memory region 7ff0b0d4b000 (8388608 bytes)
*** Preparing to test memory region 7ff0b154c000 (8388608 bytes)
*** Preparing to test memory region 7ff0b1d4d000 (8388608 bytes)
*** Preparing to test memory region 7ff0b254e000 (8388608 bytes)
*** Preparing to test memory region 7ff0b2d4f000 (8388608 bytes)
*** Preparing to test memory region 7ff0b3550000 (8388608 bytes)
*** Preparing to test memory region 7ff0b3d51000 (8388608 bytes)
*** Preparing to test memory region 7ff0b4552000 (8388608 bytes)
*** Preparing to test memory region 7ff0b4d53000 (8388608 bytes)
*** Preparing to test memory region 7ff0b5554000 (8388608 bytes)
*** Preparing to test memory region 7ff0b5d55000 (8388608 bytes)
*** Preparing to test memory region 7ff0b6556000 (8388608 bytes)
*** Preparing to test memory region 7ff0b6d57000 (8388608 bytes)
*** Preparing to test memory region 7ff0b7558000 (8388608 bytes)
*** Preparing to test memory region 7ff0b7d59000 (8388608 bytes)
*** Preparing to test memory region 7ff0b855a000 (8388608 bytes)
*** Preparing to test memory region 7ff0b8d5b000 (8388608 bytes)
*** Preparing to test memory region 7ff0b955c000 (8388608 bytes)
*** Preparing to test memory region 7ff0b9d5d000 (8388608 bytes)
*** Preparing to test memory region 7ff0ba55e000 (8388608 bytes)
*** Preparing to test memory region 7ff0bad5f000 (8388608 bytes)
*** Preparing to test memory region 7ff0bb560000 (8388608 bytes)
*** Preparing to test memory region 7ff0bbd61000 (8388608 bytes)
*** Preparing to test memory region 7ff0bc562000 (8388608 bytes)
*** Preparing to test memory region 7ff0bcd63000 (8388608 bytes)
*** Preparing to test memory region 7ff0bd564000 (8388608 bytes)
*** Preparing to test memory region 7ff0bdffc000 (4096 bytes)
*** Preparing to test memory region 7ff0bdffe000 (8388608 bytes)
*** Preparing to test memory region 7ff0be7ff000 (8388608 bytes)
*** Preparing to test memory region 7ff0bf000000 (12582912 bytes)
*** Preparing to test memory region 7ff0c011d000 (16384 bytes)
*** Preparing to test memory region 7ff0c033a000 (16384 bytes)
*** Preparing to test memory region 7ff0c0a5b000 (20480 bytes)
*** Preparing to test memory region 7ff0c0a60000 (12288 bytes)
*** Preparing to test memory region 7ff0c0a65000 (4096 bytes)
.O.O.O.O.O.O.
```

**Comment by mnunberg** [comment]
With the _new_ code? hrm.

**Comment by mnunberg** [comment]
btw, did you wipe the rdb file between runs?

**Comment by rthadr** [comment]
Yes, sorry, I should have comment on the PR.  This is on a fresh docker container.  [Here](https://pastebin.com/KaxRdpxn) is the list of numbers I'm indexing.  I'm writing them as fast as our database can send them.

**Comment by mnunberg** [comment]
Thanks for the list of numbers. I'll see what's causing it tomorrow - it may be a latent bug not even introduced in this commit.

**Comment by rthadr** [comment]
Cool, I need to sign off as well.  I believe indexing this list worked fine on master.

FYI this is only part of the list I'll be indexing, I just sent you the list up to the point it crashed.

**Comment by mnunberg** [comment]
Should work now.

**Comment by rthadr** [comment]
Awesome, I'll test it out.

**Comment by mnunberg** [comment]
@dvirsky btw I've been thinking if we shouldn't just (optionally) encode numeric fields as deltas anyway. This is especially useful for timestamps where often the "real" value is much smaller, and we can reasonably expect most values to be in the 20th/21st century etc.

**Comment by rthadr** [comment]
@mnunberg works for me!  Thank you.

**Comment by dvirsky** [comment]
This is now in master and will be in release 0.20 in a few minutes. @rthadr the implementation has been rewritten, could you please verify that it still works correctly?

**Comment by rthadr** [comment]
@dvirsky - Just pulled the latest from the numcode2.0 branch and everything is working as expected.  Thanks!

**Comment by dvirsky** [comment]
@rthadr thanks, No need for a special branch anymore, it's in master and released as version 0.20.0 now.
